tiny-event-sourcing-lib/src/main/kotlin/ru/quipy/saga/aggregate/logic/SagaStepAggregateState.kt

```diff
@@ -0,0 +1,101 @@
+import ru.quipy.domain.AggregateState
+import ru.quipy.saga.SagaStep
+import ru.quipy.saga.aggregate.api.*
+import java.util.UUID
+
+class SagaStepAggregateState : AggregateState<UUID, SagaStepAggregate> {
+    private lateinit var sagaName: String
+    private lateinit var sagaInstanceId: UUID
+    private var sagaSteps = mutableListOf<UUID>()
+    private var processedSagaSteps = mutableListOf<UUID>()
+    override fun getId() = sagaInstanceId
+
+    fun launchSagaStep(sagaStep: SagaStep): SagaStepLaunchedEvent {
+        return SagaStepLaunchedEvent(
+            sagaStep.sagaName,
+            sagaStep.stepName,
+            sagaStep.sagaStepId,
+            sagaStep.sagaInstanceId,
+            sagaStep.prevSteps
+        )
+    }
+
+    fun initiateSagaStep(sagaStep: SagaStep): SagaStepInitiatedEvent {
+        if (sagaSteps.contains(sagaStep.sagaStepId)) {
+            throw IllegalStateException("Duplicate step: $sagaStep")
+        }
+
+        return SagaStepInitiatedEvent(
+            sagaStep.sagaName,
+            sagaStep.stepName,
+            sagaStep.sagaStepId,
+            sagaStep.sagaInstanceId,
+            sagaStep.prevSteps
+        )
+    }
+
+    fun processSagaStep(sagaStep: SagaStep, eventName: String): SagaStepProcessedEvent {
+        if (processedSagaSteps.contains(sagaStep.sagaStepId)) {
+            throw IllegalStateException("Duplicate step: $sagaStep")
+        }
+
+        return SagaStepProcessedEvent(
+            sagaStep.sagaName,
+            sagaStep.stepName,
+            sagaStep.sagaStepId,
+            sagaStep.sagaInstanceId,
+            sagaStep.prevSteps,
+            eventName
+        )
+    }
+
+    fun containsProcessedSagaStep(sagaStepId: UUID): Boolean {
+        return processedSagaSteps.contains(sagaStepId)
+    }
+
+    fun processMinSaga(
+        correlationId: UUID,
+        currentEventId: UUID,
+        eventName: String,
+        causationId: UUID? = null
+    ): MinSagaProcessedEvent {
+        if (processedSagaSteps.contains(currentEventId)) {
+            throw IllegalStateException("Duplicate step: $correlationId")
+        }
+
+        return MinSagaProcessedEvent(
+            correlationId,
+            currentEventId,
+            causationId,
+            eventName
+        )
+    }
+
+    @StateTransitionFunc
+    fun launchSagaStep(sagaStepEvent: SagaStepLaunchedEvent) {
+        sagaName = sagaStepEvent.sagaName
+        sagaInstanceId = sagaStepEvent.sagaInstanceId
+        sagaSteps.add(sagaStepEvent.sagaStepId)
+    }
+
+    @StateTransitionFunc
+    fun initiateSagaStep(sagaStepEvent: SagaStepInitiatedEvent) {
+        sagaSteps.add(sagaStepEvent.sagaStepId)
+    }
+
+    @StateTransitionFunc
+    fun processSagaStep(sagaStepEvent: SagaStepProcessedEvent) {
+        processedSagaSteps.add(sagaStepEvent.sagaStepId)
+    }
+
+    @StateTransitionFunc
+    fun processMinSaga(sagaEvent: MinSagaProcessedEvent) {
+        if (!this::sagaInstanceId.isInitialized) {
```

Comment:
why this isInitialized?

Reply:
Unlike the "extended" Saga, which is created using `SagaManager`, the "default" Saga does not have a separation of events into "Initiated" - those that have not yet been written to the database, and "Processed" - those that have already been successfully processed and written to the database. In the "default" Saga there are only "Processed" events. And this check is needed so that only the first event initializes `sagaInstanceId` in the local Saga aggregate, because one event type is used for both the creation of Saga and its continuation in this aggregate

We can create an additional event like `firstProcessDefaultSaga` to avoid this check